src/server/server_family.cc

```diff
@@ -3179,6 +3206,39 @@ void ServerFamily::Module(CmdArgList args, const CommandContext& cmd_cntx) {
   rb->SendLong(20'000);  // we target v2
 }
 
+void ServerFamily::ClientPauseCmd(CmdArgList args, SinkReplyBuilder* builder,
+                                  ConnectionContext* cntx) {
+  CmdArgParser parser(args);
+  auto listeners = GetNonPriviligedListeners();
+
+  auto timeout = parser.Next<uint64_t>();
+  ClientPause pause_state = ClientPause::ALL;
+  if (parser.HasNext()) {
+    pause_state = parser.MapNext("WRITE", ClientPause::WRITE, "ALL", ClientPause::ALL);
+  }
+  if (auto err = parser.Error(); err) {
+    return builder->SendError(err->MakeReply());
+  }
+
+  const auto timeout_ms = timeout * 1ms;
+  auto is_pause_in_progress = [this, end_time = chrono::steady_clock::now() + timeout_ms] {
+    return ServerState::tlocal()->gstate() != GlobalState::SHUTTING_DOWN &&
+           chrono::steady_clock::now() < end_time && client_pause_.load(std::memory_order_relaxed);
+  };
+
+  std::unique_lock lk(client_pause_mu_);
+  if (auto pause_fb_opt =
+          Pause(listeners, cntx->ns, cntx->conn(), pause_state, std::move(is_pause_in_progress));
+      pause_fb_opt) {
+    client_pause_fb_.JoinIfNeeded();
```

Comment:
here you wait for the previous pause command to finish right? this will make this pause execution to hang until the previous paused finished.

Reply:
also taking client_pause_mu_ above and waiting here means that you will not be able to run client unpause untill this wait is done

Reply:
Assume two client pause write commands one after the other. I did the blocking on purpose because I thought that it's illegal/we must block the second `client pause 10000 write`. 

However, I was wrong! I tried with Valkey as well to confirm this behaviour.

Good catch, I did break the semantics here. And since we did not have a test for this I will also add one so we don't break it in the future.

It's funny, cause I experimented with `unpause` with valkey before I implemented this and I still missed this behavour :facepalm:

Reply:
fixed